File: core/src/main/scala/net/kemitix/thorp/core/ConfigOption.scala
package net.kemitix.thorp.core

import java.nio.file.Path

import net.kemitix.thorp.domain
import net.kemitix.thorp.domain.{Config, RemoteKey}

sealed trait ConfigOption {
  def update(config: Config): Config
}

object ConfigOption {
  case object Version extends ConfigOption {
    override def update(config: Config): Config = config
  }
  case object BatchMode extends ConfigOption {
    override def update(config: Config): Config = config.copy(batchMode = true)
  }
  case class Source(path: Path) extends ConfigOption {
    override def update(config: Config): Config = config.copy(source = path)
  }
  case class Bucket(name: String) extends ConfigOption {
    override def update(config: Config): Config = config.copy(bucket = domain.Bucket(name))
  }
  case class Prefix(path: String) extends ConfigOption {
    override def update(config: Config): Config = config.copy(prefix = RemoteKey(path))
  }
  case class Include(pattern: String) extends ConfigOption {
    override def update(config: Config): Config = config.copy(filters = domain.Filter.Include(pattern) :: config.filters)
  }
  case class Exclude(pattern: String) extends ConfigOption {
    override def update(config: Config): Config = config.copy(filters = domain.Filter.Exclude(pattern) :: config.filters)
  }
  case class Debug() extends ConfigOption {
    override def update(config: Config): Config = config.copy(debug = true)
  }
  case object IgnoreUserOptions extends ConfigOption {
    override def update(config: Config): Config = config
  }
  case object IgnoreGlobalOptions extends ConfigOption {
    override def update(config: Config): Config = config
  }

}

File: config/src/main/java/net/kemitix/thorp/config/ConfigOption.java
package net.kemitix.thorp.config;

import lombok.AccessLevel;
import lombok.RequiredArgsConstructor;
import net.kemitix.mon.TypeAlias;
import net.kemitix.thorp.domain.Filter;
import net.kemitix.thorp.domain.RemoteKey;

import java.nio.file.Path;
import java.util.ArrayList;
import java.util.List;

public interface ConfigOption {
    Configuration update(Configuration config);

    static ConfigOption source(Path path) {
        return new Source(path);
    }
    class Source extends TypeAlias<Path> implements ConfigOption {
        private Source(Path value) {
            super(value);
        }
        @Override
        public Configuration update(Configuration config) {
            return config.withSources(config.sources.append(getValue()));
        }
        public Path path() {
            return getValue();
        }
    }

    static ConfigOption bucket(String name) {
        return new Bucket(name);
    }
    class Bucket extends TypeAlias<String> implements ConfigOption {
        private Bucket(String value) {
            super(value);
        }
        @Override
        public Configuration update(Configuration config) {
            return config.withBucket(
                    net.kemitix.thorp.domain.Bucket.named(getValue()));
        }
    }

    static ConfigOption prefix(String path) {
        return new Prefix(path);
    }
    class Prefix extends TypeAlias<String> implements ConfigOption {
        private Prefix(String value) {
            super(value);
        }
        @Override
        public Configuration update(Configuration config) {
            return config.withPrefix(RemoteKey.create(getValue()));
        }
    }

    static ConfigOption include(String pattern) {
        return new Include(pattern);
    }
    class Include extends TypeAlias<String> implements ConfigOption {
        private Include(String value) {
            super(value);
        }
        @Override
        public Configuration update(Configuration config) {
            List<Filter> filters = new ArrayList<>(config.filters);
            filters.add(net.kemitix.thorp.domain.Filter.include(getValue()));
            return config.withFilters(filters);
        }
    }

    static ConfigOption exclude(String pattern) {
        return new Exclude(pattern);
    }
    class Exclude extends TypeAlias<String> implements ConfigOption {
        private Exclude(String value) {
            super(value);
        }
        @Override
        public Configuration update(Configuration config) {
            List<Filter> filters = new ArrayList<>(config.filters);
            filters.add(net.kemitix.thorp.domain.Filter.exclude(getValue()));
            return config.withFilters(filters);
        }
    }

    static ConfigOption debug() {
        return new Debug();
    }
    class Debug implements ConfigOption {
        @Override
        public Configuration update(Configuration config) {
            return config.withDebug(true);
        }
        @Override
        public String toString() {
            return "Debug";
        }
    }

    static ConfigOption batchMode() {
        return new BatchMode();
    }
    class BatchMode implements ConfigOption {
        @Override
        public Configuration update(Configuration config) {
            return config.withDebug(true);
        }

        @Override
        public String toString() {
            return "BatchMode";
        }
    }

    static ConfigOption version() {
        return new Version();
    }
    class Version implements ConfigOption {
        @Override
        public Configuration update(Configuration config) {
            return config;
        }
        @Override
        public String toString() {
            return "Version";
        }
    }

    static ConfigOption ignoreUserOptions() {
        return new IgnoreUserOptions();
    }
    class IgnoreUserOptions implements ConfigOption {
        @Override
        public Configuration update(Configuration config) {
            return config;
        }
        @Override
        public String toString() {
            return "Ignore User Options";
        }
    }

    static ConfigOption ignoreGlobalOptions() {
        return new IgnoreGlobalOptions();
    }
    class IgnoreGlobalOptions implements ConfigOption {
        @Override
        public Configuration update(Configuration config) {
            return config;
        }

        @Override
        public String toString() {
            return "Ignore Global Options";
        }
    }

    static ConfigOption parallel(int factor) {
        return new Parallel(factor);
    }
    @RequiredArgsConstructor(access = AccessLevel.PRIVATE)
    class Parallel implements ConfigOption {
        public final int factor;
        @Override
        public Configuration update(Configuration config) {
            return config.withParallel(factor);
        }
        @Override
        public String toString() {
            return "Parallel: " + factor;
        }
    }
}
